### PR TITLE
Added Vite/TypeScript/React to extensions, created papi.d.ts, shared types between extensions and main

### PR DESCRIPTION
**Pull request by tjcouch-sil**
Relates to https://github.com/paranext/paranext-extension-template/issues/1

This change is [Reviewable](https://reviewable.io/reviews/paranext/paranext-core/161)

### CODE REVIEW

File: extensions/lib/hello-someone/hello-someone.web-view.ejs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+          const name = greetingsInput.value;
+          const setGreetingsInput = document.getElementById('set-greetings-input');
+          const success = await greetingsDataProvider.set(
+            name.toLowerCase(),
+            setGreetingsInput.value,
+          );
+          const helloSomeoneOutput = document.getElementById('greetings-button-output');
+          helloSomeoneOutput.innerHTML = `${
+            success ? 'Successfully' : 'Unsuccessfully'
+          } set ${name}'s greeting!`;
```
[the remark this answers is not in the document]
Comment (tjcouch-sil) [reply extensions/lib/hello-someone/hello-someone.web-view.ejs:67]: Fixed

File: extensions/lib/hello-someone/hello-someone.web-view.ejs [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+      async function setupWebView() {
+        const greetingsDataProvider = await dataProviderPromise;
+
+        // Attach handler for greetings-button
+        const greetingsButton = document.getElementById('greetings-button');
+        greetingsButton.addEventListener('click', async () => {
+          const greetingsInput = document.getElementById('greetings-input');
+          const greeting = await greetingsDataProvider.get(greetingsInput.value.toLowerCase());
+          const helloSomeoneOutput = document.getElementById('greetings-button-output');
+          helloSomeoneOutput.innerHTML = greeting;
```
[the remark this answers is not in the document]
Comment (tjcouch-sil) [reply extensions/lib/hello-someone/hello-someone.web-view.ejs:41]: Fixed

**Review by lyonsil — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/161)* status: 0 of 52 files reviewed, 3 unresolved discussions (waiting on @github-code-scanning[bot] and @tjcouch-sil)

___
*[`src/shared/services/network-object.service.ts` line 458 at r3](https://reviewable.io/reviews/paranext/paranext-core/161#-NTzgljfDwKGlT8q9T5H:-NTzgljfDwKGlT8q9T5I:blh2xb1) ([raw file](https://github.com/paranext/paranext-core/blob/dc13614c5376e61ae0111bd0d2eba5a004d382e5/src/shared/services/network-object.service.ts#L458)):*
> ```XML
>     // Remove onDidDispose from objectToShare because the NetworkableObject<T> had its onDidDispose
>     // removed in overrideOnDidDispose
>     return objectToShare as Omit<
> ```

The object returned from `set` should have `onDidDispose` on it.  We don't allow the input object to contain one because we add one along the way.  `DisposableNetworkObject` includes an `onDidDispose` on it, so by omitting it we're breaking the contract.

**Review by tjcouch-sil — COMMENTED** [review]
> Dismissed @github-code-scanning[bot] from 2 discussions.
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/161)* status: 0 of 57 files reviewed, 1 unresolved discussion (waiting on @github-code-scanning[bot] and @lyonsil)

___
*[`src/shared/services/network-object.service.ts` line 458 at r3](https://reviewable.io/reviews/paranext/paranext-core/161#-NTzgljfDwKGlT8q9T5H:-NTzpcAi1Lz5r55vqXOv:b-t2s56d) ([raw file](https://github.com/paranext/paranext-core/blob/dc13614c5376e61ae0111bd0d2eba5a004d382e5/src/shared/services/network-object.service.ts#L458)):*
<details><summary><i>Previously, lyonsil wrote…</i></summary><blockquote>

The object returned from `set` should have `onDidDispose` on it.  We don't allow the input object to contain one because we add one along the way.  `DisposableNetworkObject` includes an `onDidDispose` on it, so by omitting it we're breaking the contract.
</blockquote></details>

I was just fixing the cyclical type problem we found the other day in this case. Basically the `papi.d.ts` file was generating with this `set` function returning `Promise<never>` because it couldn't figure out this double-defined `onDidDispose`. So I just removed it from the type that forces it to `undefined` before asserting to `DisposableNetworkObject` so it can be the correct type :) Updated the comment - hopefully the comment explains it better now

**Review by lyonsil — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/161)* status: 0 of 57 files reviewed, 1 unresolved discussion (waiting on @github-code-scanning[bot] and @tjcouch-sil)

___
*[`src/shared/services/network-object.service.ts` line 458 at r3](https://reviewable.io/reviews/paranext/paranext-core/161#-NTzgljfDwKGlT8q9T5H:-NTzumwK5J27RraIbtys:bma1jes) ([raw file](https://github.com/paranext/paranext-core/blob/dc13614c5376e61ae0111bd0d2eba5a004d382e5/src/shared/services/network-object.service.ts#L458)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

I was just fixing the cyclical type problem we found the other day in this case. Basically the `papi.d.ts` file was generating with this `set` function returning `Promise<never>` because it couldn't figure out this double-defined `onDidDispose`. So I just removed it from the type that forces it to `undefined` before asserting to `DisposableNetworkObject` so it can be the correct type :) Updated the comment - hopefully the comment explains it better now
</blockquote></details>

My concern is that adding `Omit` as it it in the PR doesn't seem right.  `set` returns a `DisposableNetworkObject`, and that requires `OnDidDispose`.  If we omit that event, we aren't actually conforming to the output type.  `set` is effectively transforming a `NetworkableObject` into a `DisposableNetworkObject`.  It seems better to go back to casting through `unknown` than to say we're removing a required property.

**Review by lyonsil — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/161)* status: 0 of 57 files reviewed, 1 unresolved discussion (waiting on @github-code-scanning[bot] and @tjcouch-sil)

___
*[`src/shared/services/network-object.service.ts` line 458 at r3](https://reviewable.io/reviews/paranext/paranext-core/161#-NTzgljfDwKGlT8q9T5H:-NTzvwr91GrxNHbcvnY1:b7qvmxi) ([raw file](https://github.com/paranext/paranext-core/blob/dc13614c5376e61ae0111bd0d2eba5a004d382e5/src/shared/services/network-object.service.ts#L458)):*
<details><summary><i>Previously, lyonsil wrote…</i></summary><blockquote>

My concern is that adding `Omit` as it it in the PR doesn't seem right.  `set` returns a `DisposableNetworkObject`, and that requires `OnDidDispose`.  If we omit that event, we aren't actually conforming to the output type.  `set` is effectively transforming a `NetworkableObject` into a `DisposableNetworkObject`.  It seems better to go back to casting through `unknown` than to say we're removing a required property.
</blockquote></details>

Although maybe this is just all a type fallacy and the object itself doesn't actually have `onDidDispose` removed by `Omit`, and it's just playing typing games?  Keeping track of what is "real" and what is "fake" is not something I'm used to doing.

**Review by lyonsil — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/161)* status: 0 of 57 files reviewed, all discussions resolved (waiting on @github-code-scanning[bot])

___
*[`src/shared/services/network-object.service.ts` line 458 at r3](https://reviewable.io/reviews/paranext/paranext-core/161#-NTzgljfDwKGlT8q9T5H:-NU20BFA9p-7mtpg7E9q:becrgme) ([raw file](https://github.com/paranext/paranext-core/blob/dc13614c5376e61ae0111bd0d2eba5a004d382e5/src/shared/services/network-object.service.ts#L458)):*
<details><summary><i>Previously, lyonsil wrote…</i></summary><blockquote>

Although maybe this is just all a type fallacy and the object itself doesn't actually have `onDidDispose` removed by `Omit`, and it's just playing typing games?  Keeping track of what is "real" and what is "fake" is not something I'm used to doing.
</blockquote></details>

Ok - I tried this out in the debugger on my own, and indeed I see this is just in "type space", not "object space."

**Review by lyonsil — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/161)* status: 0 of 57 files reviewed, all discussions resolved (waiting on @github-code-scanning[bot])

**Review by lyonsil — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/161)* status: 0 of 57 files reviewed, 1 unresolved discussion (waiting on @github-code-scanning[bot] and @tjcouch-sil)

___
*[`src/shared/services/data-provider.service.ts` line 319 at r4](https://reviewable.io/reviews/paranext/paranext-core/161#-NU22f3B-Em_quQ9JU38:-NU22f3B-Em_quQ9JU39:b-f0do2o) ([raw file](https://github.com/paranext/paranext-core/blob/070eb9139c788243b3c5964929f587d8eb165070/src/shared/services/data-provider.service.ts#L319)):*
> ```XML
>  */
> // eslint-disable-next-line @typescript-eslint/no-explicit-any
> async function get<T extends IDataProvider<any, any, any>>(
> ```

Doesn't this cause a problem in "type space" since `T` applies transitively down to `createLocalDataProviderToProxy`?  That is, since T now includes `onDidDispose` with these changes, the return type of `createLocalDataProviderToProxy` is `Partial<NetworkableObject<DataProvider<any, any, any>>>`.  The definitions of `onDidDispose` for `NetworkableObject` and `DataProvider` now collide which seems like the same problem you ran into in network-object.service.ts.  I think you want to leave T as `IDataProvider` here, not `DataProvider`.

**Review by lyonsil — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/161)* status: 0 of 57 files reviewed, 1 unresolved discussion (waiting on @github-code-scanning[bot] and @tjcouch-sil)

___
*[`src/shared/services/data-provider.service.ts` line 319 at r4](https://reviewable.io/reviews/paranext/paranext-core/161#-NU240iT8q9uxldj7-qW:-NU240iT8q9uxldj7-qX:b-dt2x66) ([raw file](https://github.com/paranext/paranext-core/blob/d265dc7cf1435c43eda8512db319b9722b73b10e/src/shared/services/data-provider.service.ts#L319)):*
> ```XML
>  */
> // eslint-disable-next-line @typescript-eslint/no-explicit-any
> async function get<T extends DataProvider<any, any, any>>(
> ```

Doesn't this cause a problem in "type space" since T applies transitively down to createLocalDataProviderToProxy?  That is, since T now includes onDidDispose with these changes, the return type of createLocalDataProviderToProxy is Partial<NetworkableObject<DataProvider<any, any, any>>>.  The definitions of onDidDispose for NetworkableObject and DataProvider now collide which seems like the same problem you ran into in network-object.service.ts.  I think you want to leave T as IDataProvider here, not DataProvider.

**Review by tjcouch-sil — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/161)* status: 0 of 58 files reviewed, 1 unresolved discussion (waiting on @github-code-scanning[bot] and @lyonsil)

___
*[`src/shared/services/data-provider.service.ts` line 319 at r4](https://reviewable.io/reviews/paranext/paranext-core/161#-NU240iT8q9uxldj7-qW:-NU2vo761ivZHuE2SdFH:b-hwhpmk) ([raw file](https://github.com/paranext/paranext-core/blob/d265dc7cf1435c43eda8512db319b9722b73b10e/src/shared/services/data-provider.service.ts#L319)):*
<details><summary><i>Previously, lyonsil wrote…</i></summary><blockquote>

Doesn't this cause a problem in "type space" since T applies transitively down to createLocalDataProviderToProxy?  That is, since T now includes onDidDispose with these changes, the return type of createLocalDataProviderToProxy is Partial<NetworkableObject<DataProvider<any, any, any>>>.  The definitions of onDidDispose for NetworkableObject and DataProvider now collide which seems like the same problem you ran into in network-object.service.ts.  I think you want to leave T as IDataProvider here, not DataProvider.
</blockquote></details>

Looks like `dataProviderService.get` with `T extends DataProvider<any, any, any>` does kinda implicitly violate the type rules we have set up as you mention, but it doesn't cause any issues. Probably because the type issues we're seeing are problems with functions passed around internally in the JS, not anything in the TS. So I'd say it might be best to leave it as the `dataProviderService.get` returns a `DataProvider`, not just an `IDataProvider`.
___
*[`src/shared/services/network-object.service.ts` line 458 at r3](https://reviewable.io/reviews/paranext/paranext-core/161#-NTzgljfDwKGlT8q9T5H:-NU2w0XI-H7Jpiyofl21:b55384y) ([raw file](https://github.com/paranext/paranext-core/blob/dc13614c5376e61ae0111bd0d2eba5a004d382e5/src/shared/services/network-object.service.ts#L458)):*
<details><summary><i>Previously, lyonsil wrote…</i></summary><blockquote>

Ok - I tried this out in the debugger on my own, and indeed I see this is just in "type space", not "object space."
</blockquote></details>

Right, `as` is a type assertion, meaning it is just telling TypeScript hints as to things we know that it can't figure out on its own. It doesn't actually do anything to the variable itself. Definitely tricky to get used to!

**Review by lyonsil — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/161)* status: 0 of 58 files reviewed, 1 unresolved discussion (waiting on @github-code-scanning[bot] and @tjcouch-sil)

___
*[`src/shared/services/data-provider.service.ts` line 319 at r4](https://reviewable.io/reviews/paranext/paranext-core/161#-NU240iT8q9uxldj7-qW:-NU2zYFH4Rq9r6SwxTi1:b-4nb83o) ([raw file](https://github.com/paranext/paranext-core/blob/d265dc7cf1435c43eda8512db319b9722b73b10e/src/shared/services/data-provider.service.ts#L319)):*
<details><summary><i>Previously, tjcouch-sil (TJ Couch) wrote…</i></summary><blockquote>

Looks like `dataProviderService.get` with `T extends DataProvider<any, any, any>` does kinda implicitly violate the type rules we have set up as you mention, but it doesn't cause any issues. Probably because the type issues we're seeing are problems with functions passed around internally in the JS, not anything in the TS. So I'd say it might be best to leave it as the `dataProviderService.get` returns a `DataProvider`, not just an `IDataProvider`.
</blockquote></details>

I'm concerned about this change because it's very similar to kind of problem that I saw, then we didn't see while reviewing together, then seemed to cause a problem with later workflows in network-object.service.ts in our other discussion in this PR.  Is there a reason to change the signature here?  I think it was correct.  As specified now, `createLocalDataProviderToProxy` returns a nonsense type.

In both the old and the new signature, `dataProviderService.get` returns a `DataProvider`.  Now, though, the definition of `T` changes which impacts the definition of `T` implicitly for `createLocalDataProviderToProxy`.

**Review by tjcouch-sil — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/161)* status: 0 of 59 files reviewed, 1 unresolved discussion (waiting on @github-code-scanning[bot] and @lyonsil)

___
*[`src/shared/services/data-provider.service.ts` line 319 at r4](https://reviewable.io/reviews/paranext/paranext-core/161#-NU240iT8q9uxldj7-qW:-NU3ZxnL0E089JYS6NCY:b3r508d) ([raw file](https://github.com/paranext/paranext-core/blob/d265dc7cf1435c43eda8512db319b9722b73b10e/src/shared/services/data-provider.service.ts#L319)):*
<details><summary><i>Previously, lyonsil wrote…</i></summary><blockquote>

I'm concerned about this change because it's very similar to kind of problem that I saw, then we didn't see while reviewing together, then seemed to cause a problem with later workflows in network-object.service.ts in our other discussion in this PR.  Is there a reason to change the signature here?  I think it was correct.  As specified now, `createLocalDataProviderToProxy` returns a nonsense type.

In both the old and the new signature, `dataProviderService.get` returns a `DataProvider`.  Now, though, the definition of `T` changes which impacts the definition of `T` implicitly for `createLocalDataProviderToProxy`.
</blockquote></details>

We discussed and changed the way we did it to use `Omit<T, 'onDidDispose'>` to avoid these type paradoxes.

**Review by lyonsil — DISMISSED** [review]
> Reviewed 1 of 47 files at r1, 3 of 3 files at r6, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/161)* status: 5 of 59 files reviewed, all discussions resolved (waiting on @github-code-scanning[bot])

**Review by lyonsil — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/paranext/paranext-core/161#-:-NU3otSUEGcrsVctLckq:bnfp4nl)* status: 5 of 60 files reviewed, all discussions resolved (waiting on @github-code-scanning[bot])

**Pull request merged**